File: scripts/speculos-ui.ts
#!/usr/bin/env bun

import { Socket } from "node:net";

type ScreenEvent = {
  text?: string;
  x?: number;
  y?: number;
  clear?: boolean;
};

type ScreenResponse = {
  events?: ScreenEvent[];
};

const apiPort = process.env.SPECULOS_API_PORT ?? "5001";
const apiBase = process.env.SPECULOS_API_URL ?? `http://127.0.0.1:${apiPort}`;
const automationPort = Number(process.env.SPECULOS_AUTOMATION_PORT ?? "41000");
const buttonPort = Number(process.env.SPECULOS_BUTTON_PORT ?? "42000");
const automationHost = process.env.SPECULOS_AUTOMATION_HOST ?? "127.0.0.1";
const buttonHost = process.env.SPECULOS_BUTTON_HOST ?? "127.0.0.1";
const [command = "screen"] = process.argv.slice(2);

switch (command) {
  case "screen":
    await printCurrentScreen();
    break;
  case "events":
    await streamEvents();
    break;
  case "left":
  case "right":
  case "both":
    await pressButton(command);
    await sleep(150);
    await printCurrentScreen();
    break;
  case "clear-events":
    await clearEvents();
    break;
  default:
    printUsage();
    process.exitCode = 1;
    break;
}

async function printCurrentScreen(): Promise<void> {
  try {
    const response = await fetchJson<ScreenResponse>(`${apiBase}/events?currentscreenonly=true`);
    const lines = normalizeTexts(response.events ?? []);

    console.log(`API: ${apiBase}`);
    if (lines.length === 0) {
      console.log("(no text events)");
      return;
    }

    for (const line of lines) {
      console.log(line);
    }
  } catch {
    const lines = await collectAutomationEvents(250);
    console.log(`Automation: ${automationHost}:${automationPort}`);
    if (lines.length === 0) {
      console.log("(no fresh text events; use `bun run speculos:events` while pressing buttons)");
      return;
    }
    for (const line of lines) {
      console.log(line);
    }
  }
}

async function streamEvents(): Promise<void> {
  try {
    const response = await fetch(`${apiBase}/events?stream=true`);
    if (!response.ok || !response.body) {
      throw new Error(`Failed to open event stream: ${response.status} ${response.statusText}`);
    }

    console.error(`Streaming events from ${apiBase}/events?stream=true`);
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    for (;;) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }

      buffer += decoder.decode(value, { stream: true });
      for (;;) {
        const marker = buffer.indexOf("\n\n");
        if (marker === -1) {
          break;
        }

        const chunk = buffer.slice(0, marker);
        buffer = buffer.slice(marker + 2);

        for (const line of chunk.split("\n")) {
          if (!line.startsWith("data: ")) {
            continue;
          }
          const event = JSON.parse(line.slice(6)) as ScreenEvent;
          if (event.clear) {
            console.log("--- clear ---");
            continue;
          }
          if (event.text) {
            console.log(event.text);
          }
        }
      }
    }
  } catch {
    console.error(`Streaming events from ${automationHost}:${automationPort}`);
    await streamAutomationEvents();
  }
}

async function pressButton(button: "left" | "right" | "both"): Promise<void> {
  try {
    const response = await fetch(`${apiBase}/button/${button}`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ action: "press-and-release" }),
    });

    if (!response.ok) {
      throw new Error(`Failed to press ${button}: ${response.status} ${response.statusText}`);
    }
  } catch {
    await sendButtonSequence(button);
  }
}

async function clearEvents(): Promise<void> {
  const response = await fetch(`${apiBase}/events`, { method: "DELETE" });
  if (!response.ok) {
    throw new Error(`Failed to clear events: ${response.status} ${response.statusText}`);
  }
}

async function fetchJson<T>(url: string): Promise<T> {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Request failed: ${response.status} ${response.statusText}`);
  }
  return (await response.json()) as T;
}

function normalizeTexts(events: ScreenEvent[]): string[] {
  const seen = new Set<string>();
  const lines: string[] = [];

  for (const event of events) {
    const text = event.text?.trim();
    if (!text || seen.has(text)) {
      continue;
    }
    seen.add(text);
    lines.push(text);
  }

  return lines;
}

async function collectAutomationEvents(durationMs: number): Promise<string[]> {
  const socket = await connectSocket(automationHost, automationPort);
  const lines: string[] = [];
  let buffer = "";

  socket.on("data", (chunk: Buffer) => {
    buffer += chunk.toString("utf8");
    for (;;) {
      const marker = buffer.indexOf("\n");
      if (marker === -1) {
        break;
      }
      const line = buffer.slice(0, marker).trim();
      buffer = buffer.slice(marker + 1);
      if (!line) {
        continue;
      }
      const event = JSON.parse(line) as ScreenEvent;
      if (event.text) {
        lines.push(event.text);
      }
    }
  });

  await sleep(durationMs);
  socket.end();
  return dedupeLines(lines);
}

async function streamAutomationEvents(): Promise<void> {
  const socket = await connectSocket(automationHost, automationPort);
  let buffer = "";

  socket.on("data", (chunk: Buffer) => {
    buffer += chunk.toString("utf8");
    for (;;) {
      const marker = buffer.indexOf("\n");
      if (marker === -1) {
        break;
      }
      const line = buffer.slice(0, marker).trim();
      buffer = buffer.slice(marker + 1);
      if (!line) {
        continue;
      }
      const event = JSON.parse(line) as ScreenEvent;
      if (event.clear) {
        console.log("--- clear ---");
      } else if (event.text) {
        console.log(event.text);
      }
    }
  });

  await new Promise<void>(() => {});
}

async function sendButtonSequence(button: "left" | "right" | "both"): Promise<void> {
  const socket = await connectSocket(buttonHost, buttonPort);
  const sequence =
    button === "left" ? "Ll" : button === "right" ? "Rr" : "LRlr";
  socket.write(sequence);
  await sleep(150);
  socket.end();
}

function connectSocket(host: string, port: number): Promise<Socket> {
  return new Promise((resolve, reject) => {
    const socket = new Socket();
    socket.once("error", reject);
    socket.connect(port, host, () => {
      socket.removeListener("error", reject);
      resolve(socket);
    });
  });
}

function dedupeLines(lines: string[]): string[] {
  const seen = new Set<string>();
  const unique: string[] = [];
  for (const line of lines) {
    if (seen.has(line)) {
      continue;
    }
    seen.add(line);
    unique.push(line);
  }
  return unique;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function printUsage(): void {
  console.error(`Usage:
  bun scripts/speculos-ui.ts screen
  bun scripts/speculos-ui.ts events
  bun scripts/speculos-ui.ts left
  bun scripts/speculos-ui.ts right
  bun scripts/speculos-ui.ts both
  bun scripts/speculos-ui.ts clear-events`);
}

File: cli/src/bin/speculos_ui.rs
use std::io::{BufRead, BufReader, Read, Write};
use std::net::TcpStream;
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{Context, Result};
use clap::{Parser, Subcommand};
use reqwest::blocking::Client;
use serde::Deserialize;
use serde_json::json;

#[derive(Parser)]
#[command(name = "speculos-ui")]
struct Cli {
    #[command(subcommand)]
    command: Command,
    #[arg(long, env = "SPECULOS_API_PORT", default_value_t = 5001)]
    api_port: u16,
    #[arg(long, env = "SPECULOS_AUTOMATION_PORT", default_value_t = 41000)]
    automation_port: u16,
    #[arg(long, env = "SPECULOS_BUTTON_PORT", default_value_t = 42000)]
    button_port: u16,
    #[arg(long, env = "SPECULOS_API_URL")]
    api_url: Option<String>,
    #[arg(long, env = "SPECULOS_AUTOMATION_HOST", default_value = "127.0.0.1")]
    automation_host: String,
    #[arg(long, env = "SPECULOS_BUTTON_HOST", default_value = "127.0.0.1")]
    button_host: String,
}

#[derive(Subcommand)]
enum Command {
    Screen,
    Events,
    Left,
    Right,
    Both,
    ClearEvents,
}

#[derive(Debug, Deserialize)]
struct ScreenEvent {
    text: Option<String>,
    clear: Option<bool>,
}

#[derive(Debug, Deserialize)]
struct ScreenResponse {
    events: Option<Vec<ScreenEvent>>,
}

fn main() -> Result<()> {
    let cli = Cli::parse();
    let api_base = cli
        .api_url
        .unwrap_or_else(|| format!("http://127.0.0.1:{}", cli.api_port));

    match cli.command {
        Command::Screen => {
            print_current_screen(&api_base, &cli.automation_host, cli.automation_port)
        }
        Command::Events => stream_events(&api_base, &cli.automation_host, cli.automation_port),
        Command::Left => press_button(
            "left",
            &api_base,
            &cli.button_host,
            cli.button_port,
            &cli.automation_host,
            cli.automation_port,
        ),
        Command::Right => press_button(
            "right",
            &api_base,
            &cli.button_host,
            cli.button_port,
            &cli.automation_host,
            cli.automation_port,
        ),
        Command::Both => press_button(
            "both",
            &api_base,
            &cli.button_host,
            cli.button_port,
            &cli.automation_host,
            cli.automation_port,
        ),
        Command::ClearEvents => clear_events(&api_base),
    }
}

fn print_current_screen(api_base: &str, automation_host: &str, automation_port: u16) -> Result<()> {
    match fetch_current_screen(api_base) {
        Ok(lines) if !lines.is_empty() => {
            println!("API: {api_base}");
            for line in lines {
                println!("{line}");
            }
            Ok(())
        }
        _ => {
            let lines = collect_automation_lines(
                automation_host,
                automation_port,
                Duration::from_millis(250),
            )?;
            println!("Automation: {automation_host}:{automation_port}");
            if lines.is_empty() {
                println!("(no text events)");
            } else {
                for line in lines {
                    println!("{line}");
                }
            }
            Ok(())
        }
    }
}

fn stream_events(api_base: &str, automation_host: &str, automation_port: u16) -> Result<()> {
    let client = Client::new();
    let response = client.get(format!("{api_base}/events?stream=true")).send();
    match response {
        Ok(mut response) => {
            let mut body = String::new();
            response.read_to_string(&mut body)?;
            for line in body.lines() {
                if let Some(data) = line.strip_prefix("data: ") {
                    let event: ScreenEvent = serde_json::from_str(data)?;
                    if event.clear.unwrap_or(false) {
                        println!("--- clear ---");
                    } else if let Some(text) = event.text {
                        println!("{text}");
                    }
                }
            }
            Ok(())
        }
        Err(_) => {
            let socket = TcpStream::connect((automation_host, automation_port))
                .with_context(|| format!("failed to connect to automation socket {automation_host}:{automation_port}"))?;
            let mut reader = BufReader::new(socket);
            loop {
                let mut line = String::new();
                reader.read_line(&mut line)?;
                if line.trim().is_empty() {
                    continue;
                }
                let event: ScreenEvent = serde_json::from_str(line.trim())?;
                if event.clear.unwrap_or(false) {
                    println!("--- clear ---");
                } else if let Some(text) = event.text {
                    println!("{text}");
                }
            }
        }
    }
}

fn press_button(
    button: &str,
    api_base: &str,
    button_host: &str,
    button_port: u16,
    automation_host: &str,
    automation_port: u16,
) -> Result<()> {
    let client = Client::new();
    if client
        .post(format!("{api_base}/button/{button}"))
        .json(&json!({ "action": "press-and-release" }))
        .send()
        .is_err()
    {
        let mut socket = TcpStream::connect((button_host, button_port)).with_context(|| {
            format!("failed to connect to button socket {button_host}:{button_port}")
        })?;
        socket.write_all(button.as_bytes())?;
        socket.write_all(b"\n")?;
    }

    thread::sleep(Duration::from_millis(150));
    print_current_screen(api_base, automation_host, automation_port)
}

fn clear_events(api_base: &str) -> Result<()> {
    Client::new()
        .delete(format!("{api_base}/events"))
        .send()?
        .error_for_status()?;
    Ok(())
}

fn fetch_current_screen(api_base: &str) -> Result<Vec<String>> {
    let response = Client::new()
        .get(format!("{api_base}/events?currentscreenonly=true"))
        .send()?
        .error_for_status()?
        .json::<ScreenResponse>()?;

    let mut lines = Vec::new();
    for event in response.events.unwrap_or_default() {
        if let Some(text) = event.text {
            if !lines.iter().any(|line| line == &text) {
                lines.push(text);
            }
        }
    }
    Ok(lines)
}

fn collect_automation_lines(host: &str, port: u16, duration: Duration) -> Result<Vec<String>> {
    let socket = TcpStream::connect((host, port))
        .with_context(|| format!("failed to connect to automation socket {host}:{port}"))?;
    socket.set_nonblocking(true)?;
    let start = Instant::now();
    let mut reader = BufReader::new(socket);
    let mut lines = Vec::new();

    while start.elapsed() < duration {
        let mut line = String::new();
        match reader.read_line(&mut line) {
            Ok(0) => thread::sleep(Duration::from_millis(10)),
            Ok(_) => {
                let event: ScreenEvent = serde_json::from_str(line.trim())?;
                if let Some(text) = event.text {
                    if !lines.iter().any(|entry| entry == &text) {
                        lines.push(text);
                    }
                }
            }
            Err(_) => thread::sleep(Duration::from_millis(10)),
        }
    }

    Ok(lines)
}
